### DISCUSSION TITLE
Make Quick-Check less strict (second try)

### DISCUSSION THREAD

**Discussion by Safihre** [sabnzbd/sabnzbd#2160] [Ideas]
Here we want to collect feedback from users in case the new and old Quick-check show a different result.

Please supply the NZB and set logging to `+Debug` in the Status and Interface settings window to get us more information.

**You can also send the logs + NZB to bugs@sabnzbd.org!**

**Comment by slaughty101** [comment]
Hey there guys,

I have a weird issue when I upgraded from b4 to b5. I'm not able to download anything (and cannot see anything in the logs either). Nothing is paused, the usenet servers connect fine, when I check them from settings, priority is set to normal, so that shouldn't be an issue.

When I click the download button just nothing happens. The download speeds sometimes show either 0MB/s or 50ish MB/s, with no difference at all, meaning there's nothing downloaded either way.

I've set the logs to debug and looked through them, but couldn't find anything peculiar in them. Maybe you can find something.

Link is this one: https://pastebin.com/C8qgWCd4

Any help would be greatly appreciated. Best wishes =)

EDIT: Oh and I didn't attach any specific NZB file cause it's happening with any NZB in the queue currently

**Reply by Safihre** [reply]
That's unfortunate, but this is not the right place. This issue is only for the Quick-Check.
Could you run SABnzbd from the command line (use Command Prompt or Powershell to start `SABnzbd-console.exe` in the SABnzbd installation folder) and the see if you see any crashes there? If so, please create a new issue.

**Reply by DavidFair** [reply]
(I know this is off-topic but I had a similar problem I fixed today). Could you check if `num_simd_decoders` is set to 0 and try setting it to 1/2 to see if it unblocks the queue.

**Reply by Safihre** [reply]
Ohhhhhhhhhhh... Of course that's it! Thanks @DavidFair !

**Comment by slaughty101** [comment]
And here's another log paste with my sab settings on the bottom (if it helps)

https://pastebin.com/8Dcayepu

**Reply by thezoggy** [reply]
`num_simd_decoders = 0` is in your config.  Config > Specials > num_simd_decoders   and set that to 2, save and restart sab.

@Safihre any idea how people are getting this set, shouldnt we safeguard against this? (set min value 1?)

**Reply by Safihre** [reply]
I forgot to force that.

**Comment by gitRedd** [comment]
Hi,
As per your  request, please see attached.
[sabnzbd.log](https://github.com/sabnzbd/sabnzbd/files/8788988/sabnzbd.log)
[The.Wedding.Party.1969.1080p.BluRay.x264-GHOULS-AsRequested.nzb.gz](https://github.com/sabnzbd/sabnzbd/files/8788989/The.Wedding.Party.1969.1080p.BluRay.x264-GHOULS-AsRequested.nzb.gz)

**Reply by Safihre** [reply]
@gitRedd There seems to be something wrong with your system. Are your disks alright? There's a whole flood of writing and reading errors.

**Reply by gitRedd** [reply]
Occasionally my external drives act up. Probably usb ports.

**Comment by puzzledsab** [comment]
I've gotten different results for several nzbs, all belonging to a set. One rar file fails using the old check but not the new. The real par2 check does not find any errors. Curiously the failing file always seems to be in a similar range of the rar sets (the last ~30%). I can debug it further if you like.

**Reply by Safihre** [reply]
I noticed this too. However, I just got an example from a user where the new method could never detect the files as broken. They have a defect in the middle of a number of files but file sizes were unchanged.
So I'll have to reinvent this, for example by running full repair if unpack fails, if the first verification was only by quick check. But this requires major work in the post processing.

**Reply by puzzledsab** [reply]
I tried running a couple of the failing nzbs again and this time they had no errors. I would normally suspect a hardware problem but my system has otherwise been very stable and I haven't seen this problem for other nzbs. It seems to me that the old check is erring on the side of caution. An alternative could be to keep the old check but rerun the test on failing files.

**Reply by Safihre** [reply]
The check itself is not the problem. I explained it in detail in the previous Discussion, but then deleted it because too much copyright material NZB's were uploaded by users (my mistake!).

In short, we want to get rid of
https://github.com/sabnzbd/sabnzbd/blob/5a7cff491d4acd82c5df2ceaecc2c7d957605f3d/sabnzbd/assembler.py#L191
Because it prevents us from writing data at the highest speeds possible, alternating for every article in the assembler between a CPU task and a disk task. The MD5 calculation takes a relatively large a amount of time that can be avoided in 99.999% of the jobs.

**Reply by puzzledsab** [reply]
Wouldn't it be better to do the writing and calculating in separate threads so that they could be done concurrently? Do the writing in the loop like before and put the data in a queue for `nzf.md5.update()` in another thread. Then wait until all the parts for the file have been calculated at the end of the `for article in nzf.decodetable:` loop if all parts are available. It could possibly be done by one of the existing data cruncher threads if you would like to limit the number of threads.

**Reply by Safihre** [reply]
I tried that, can't find the commit anymore because I deleted the branch. Unfortunately the additional thread switching barely made it any faster..

**Reply by puzzledsab** [reply]
Does the new method only calculate the hash for 16 KB of the file or do I misunderstand? If so I don't think it can be reliable. I don't think depending on rar to fail is a good idea now that some people are recommending to stop using rar.

I did some experimenting with writing data, calculating md5 and threading and I got some curious results using this script: https://gist.github.com/puzzledsab/9741092732c44f553135cacee3065861

r = number of threads
m = multiplier for block size, use something divisible by 50

With no threads, only writing took 1.85 seconds while writing + md5 took 3.2 seconds. With the following variation of threads and multipliers I got these results:

r 1, m 1 = 2.75 seconds
r 2, m 1 = 2.7 seconds
r 3, m 1 = 2.65 seconds
r 1, m 10 = 2.77 seconds
r 2, m 10 = 2.3 seconds
r 3, m 5 = 2.3 seconds
r 3, m 10 = 1.9 seconds

So the last configuration pretty much removes the overhead of doing md5 calculations measured in real time. This seems to be a particular sweet spot. I have no idea why so it may not work as well on other systems. I don't know if it's possible to use it for anything..

**Reply by Safihre** [reply]
Indeed, the new system only relies on the MD5 of the first 16K and it actually turned out pretty well of used in combination with expected file size (which is also in the par2 file). Additionally, we also fail Quick check if there are any missing articles or articles with CRC errors.
So the only reason this would fail, is if the files were broken _before_ being uploaded.

Your `m` is out of our control, it's always the article size, so around `m=1`. Because we can only feed 1 article of data at a time to the MD5.

**Reply by puzzledsab** [reply]
How about making it optional to include the full check?

Regarding m, we could collect a few blocks in a list in the loop before sending them to md5. I changed my test to collect 5 blocks before joining them and putting it in the queue. It took about 2.3 seconds with 3 threads. Increasing the number of blocks above 5 didn't improve it further.

**Reply by Safihre** [reply]
> How about making it optional to include the full check?

No this needs to work out of the box, whichever way we implement it.

> Regarding m, we could collect a few blocks in a list in the loop before sending them to md5. I changed my test to collect 5 blocks before joining them and putting it in the queue. It took about 2.3 seconds with 3 threads. Increasing the number of blocks above 5 didn't improve it further.

It's still separate blocks right, is that really beneficial in the real world? (maybe I'm misunderstanding)

**Reply by puzzledsab** [reply]
> No this needs to work out of the box, whichever way we implement it.

Isn't the problem that neither solution is perfect? One is faster in some (at least currently) edge cases, the other will catch more edge case errors. Some will probably prefer speed and others reliability.

> It's still separate blocks right, is that really beneficial in the real world? (maybe I'm misunderstanding)

It can write up to 40% faster than the old version (in my test) while still doing a full md5 check of the file.

**Reply by Safihre** [reply]
When doing MD5, even if we have a few in a list, Python still has to release and acquire the GIL for every item in the list. Which results in thread switching. 
So it's not that it can really process 5 blocks in 1 go.

**Reply by Safihre** [reply]
Just another small note that I see got lost in this ticket: using this new Quick-check method (using CRC-checks/missing-articles/MD5-of-16k/filesize) and then doing full par2 only when unrar fails is exactly what NZBGet does. I discussed this with hugbug a while ago.
This way we save a lot of CPU cycles not calculating MD5 on the 99% of jobs where the new Quick-check method is sufficient.
What would be your concerns with that approach?

**Reply by puzzledsab** [reply]
> So it's not that it can really process 5 blocks in 1 go.

I `join()`'ed them to one block before putting it in the queue in my test.

> What would be your concerns with that approach?

It only finds errors if files are compressed. There's been some campaigning to stop using rar for usenet posts so I think starting to depend on it now is not ideal. I do think that it should be able to run par2 if errors are detected in rar files, though.

**Reply by Safihre** [reply]
Right, I forgot about those kind of jobs. You're right. Wonder how NZBGet handles those..

I'm surprised there's not a huge penalty for the `join`, since it has to do all the extra memory allocations and string copying in C.

**Reply by puzzledsab** [reply]
Have you looked into this any further? I would like to make an experimental branch with configurable variations (old/new checking and threaded checking with and without joining) and try to get feedback from users who are not getting optimal performance with the current version. What do you think?

**Reply by Safihre** [reply]
Sure. The easy test user would be me and my NAS, so I can test it.
Other users are not really aware of it specifically what is limiting them. Although with 3.6.0 it will show in the Status window that "Disk" caused a delay.

**Comment by BaukeZwart** [comment]
As requested:
[Around.the.World.in.80.Days.2021.S01E04.Episode.4.1080p.AMZN.WEB-DL.DDP5.1.H.264-TEPES.nzb.gz](https://github.com/sabnzbd/sabnzbd/files/8840201/Around.the.World.in.80.Days.2021.S01E04.Episode.4.1080p.AMZN.WEB-DL.DDP5.1.H.264-TEPES.nzb.gz)

**Comment by welfordm** [comment]
Luther.S04.1080p.BluRay.x264-CADAVER

**Comment by bambury** [comment]
Hi, I just updated to 3.6.0 and I'm having a quick-check-related issue, so I thought I'd post about it here in case it's pertinent.

Essentially, the NZB contains identical files with different names (because of reasons). Sab renames fileS01E01.ext to fileS01E02.ext (which I'd rather not to), then also tries to rename fileS01E01.ext to fileS01E03.ext (and so on, I'd guess) and stops with an error because of course fileS01E01.ext doesn't exist anymore.

 Debug log extract is [here](https://0bin.net/paste/PzL+GrL4#2nQVv8WYp81coUcNhI6KiiiSOZcqO-D04fEoTAj0YNG).

NZB file is [here](https://anonfiles.com/X7X9b7oeya/The.Blacklist.S01.BDRip.x264-DEMAND_nzb).

**Reply by Safihre** [reply]
Please create a new issue for this. It's unrelated.

**Comment by Simplehorror** [comment]
[Parks.and.Recreation.S04E12.1080p.BluRay.x264-BORDURE-xpost.nzb.zip](https://github.com/sabnzbd/sabnzbd/files/9024271/Parks.and.Recreation.S04E12.1080p.BluRay.x264-BORDURE-xpost.nzb.zip)
Not sure where to find the debug logs.
